feat(server): Retry network failures on upstream requests

RaduW opened the pull request:
Implements a retry mechanism for upstream http requests that have failed due to network errors. Network errors are timeouts, payload errors, and responses with status code 502, 503 and 504. 

When Relay receives repeated errors for a configurable time defaulting to 10s, it moves into unauthenticated state. At this point, all further requests are suspended until authentication can be established again. Every successful request resets the timer.

Not all requests are retried automatically. Project config queries and Relay ID queries have special batched retry handling. Envelope and event submission is retried by default.

RaduW committed:
made requests re-sendable, added immediate request priority

RaduW committed:
first draft, retry network errors

RaduW committed:
added an interval for stopping http messages after a network error

After network errors we pause the sending of queue requests for
a period. Non queued request will still go through.

RaduW committed:
minor clenup

RaduW committed:
unified request retry mechanism

RaduW committed:
small fix

RaduW committed:
Merge branch 'master' into ref/retry-failed-http

# Conflicts:
# relay-server/src/actors/upstream.rs

RaduW committed:
Update change log

RaduW committed:
Merge branch 'master' into ref/retry-failed-http

# Conflicts:
#	CHANGELOG.md

jan-auer commented on CHANGELOG.md:
```diff
@@ -21,6 +21,7 @@
 - Experimental data scrubbing on minidumps([#682](https://github.com/getsentry/relay/pull/682))
 - Move `generate-schema` from the Relay CLI into a standalone tool. ([#739](//github.com/getsentry/relay/pull/739))
 - Move `process-event` from the Relay CLI into a standalone tool. ([#740](//github.com/getsentry/relay/pull/740))
+- Retry sending events with network errors. ([#747](//github.com/getsentry/relay/pull/747))
```
_nit_: Let's move this up into the "Features" section, since this is relevant to Managed Relays.

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
+        };
+        let queue: Option<_>;
+
+        match request.priority {
```
You can make the queue non-optional. Simply do this:

```rust
let queue = match request.priority { ... };
```

Since you return in the `Immediate` branch, it will still work.

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
+
+        if let Some(queue) = queue {
+            let name = request.priority.name();
+            push(queue, request);
```
_nit_: Please inline the match on position for readability:

```suggestion
            match position {
                EnqueuePosition::Front => queue.push_front(request),
                EnqueuePosition::Back => queue.push_back(request),
            }
```

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
+        request: UpstreamRequest,
+        send_result: Result<ClientResponse, UpstreamRequestError>,
+        ctx: &mut Context<Self>,
+    ) -> impl Future<Item = (), Error = ()> {
```
Please name the future type that is returned from this function.

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
+            request.response_sender.send(Err(err)).ok();
+        } else {
+            // reset any previously failed status
+            // TODO should we reset on success or on anything that is not a network error ?
```
This should reset on anything that is not a network error.

RaduW replied:
Ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
+    ///         - if there is no scheduled Authentication request
+    ///             - schedule authentication (with backoff)
+    ///     - if it was a non recoverable error, notify the response sender with the error
+    fn handle_http_response_status(
```
_nit_: This actually handles the complete response, so potentially `handle_http_response` is more appropriate.

RaduW replied:
ok

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -276,161 +292,226 @@ impl UpstreamRelay {
+                path: path.as_ref().to_owned(),
+                response_sender: tx,
+                build: Box::new(build),
+            },
```
_nit_: Declare in a variable for readability.

jan-auer commented on relay-server/src/actors/upstream.rs:
```diff
@@ -728,6 +826,9 @@ pub trait UpstreamQuery: Serialize {
     fn priority() -> RequestPriority {
         RequestPriority::Low
     }
+    fn retry() -> bool {
```
_nit_: Add an empty line. 

I am thinking that default behavior should be to retry all queries, and we only selectively disable where we have more appropriate logic. In fact, most queries have custom logic, but it's still "safer" to assume retries.

should retry:
- `SendOutcomes`
- `GetProjectIds`

custom logic:
- `GetProjectStates`
- `GetRelays`
- `RegisterRequest`
- `RegisterResponse`

RaduW replied:
I was thinking that it would be safer to be explicit about when there are retries since this would be a change from the previous behaviour. If you go specifically to a message and mark it as retry then presumably you know what you are doing, as opposed to have everything change and then go and mark what shouldn't change.

jan-auer replied:
Ok, this makes sense. Let's keep it `false` by default, then, or remove the default.

jan-auer reviewed (commented):
Preliminary review. Going to do another pass over the logic.

This needs E2E tests to ensure the behavior works as expected as this is a high risk change.

RaduW committed:
Added integration tests and fixed one bug

RaduW committed:
Changes from code review

RaduW committed:
removed default implementation fo UpstreamQuery::retry

Make it obvious what a message does by explicitly specifying
its retry behaviour.

jan-auer committed:
fix: Code cosmetics and send non-retry errors

jan-auer committed:
test: Add a test for buffering before authentication

jan-auer committed:
meta: Changelog
